lib/eth_web_socket/server.ex

```diff
@@ -0,0 +1,246 @@
+defmodule EthWebSocket.Server do
+  use GenServer
+  alias EthWebSocket.Client
+  @time_out 60_000
+
+  def start_link(opts \\ []) do
+    GenServer.start_link(__MODULE__, :ok, opts)
+  end
+
+  def init(:ok) do
+    {:ok,
+     %{
+       requests: 0,
+       new_heads: %{request_id: nil, sub: [], subscription_id: nil},
+       new_pending_transactions: %{request_id: nil, sub: [], subscription_id: nil}
+     }}
+  end
+
+  def start_server do
+    {:ok, pid} = start_link()
+    init_ws(pid)
+    pid
+  end
+
+  def init_ws(pid) do
+    GenServer.call(pid, {:init_ws, pid})
+  end
+
+  def request("eth_subscribe", ["newPendingTransactions"], pid) do
+    handle_sub_request(pid, ["newPendingTransactions"], :new_pending_transactions)
+  end
+
+  def request("eth_subscribe", params, pid) do
+    handle_sub_request(pid, params, :new_heads)
+  end
+
+  def request("eth_unsubscribe", params, pid) do
+    state = get_state(pid)
+
+    sub_atom = get_sub_atom(state, List.first(params), :subscription_id)
+
+    subs = state[sub_atom]
+
+    case length(subs[:sub]) do
+      n when n > 1 ->
+        GenServer.call(pid, {:unsubscribe, sub_atom})
+        {:ok, true}
+
+      1 ->
+        id = state[:requests]
+        GenServer.call(pid, :increment_request_count)
+        {:ok, body} = add_request_info("eth_unsubscribe", params, id)
+        GenServer.call(pid, {:unsubscribe, sub_atom})
+        GenServer.call(pid, {:clean_sub_id, sub_atom})
+        send(pid, {:perform_asynchronous_request, body})
+        GenServer.call(pid, :wait_for_answer, @time_out)
+
+      _ ->
+        {:ok, false}
+    end
+  end
+
+  def request(method_name, params, pid) do
+    id = get_state(pid)[:requests]
+    GenServer.call(pid, :increment_request_count)
+    {:ok, body} = add_request_info(method_name, params, id)
+    send(pid, {:perform_asynchronous_request, body})
+
+    GenServer.call(pid, :wait_for_answer, @time_out)
+  end
+
+  def handle_sub_request(pid, params, sub_atom) do
```

Comment:
Should this function be accessible from outside this module?

Reply:
I noticed that and have already made the changes, but I forgot to commit them 😅